File: android/app/src/main/java/com/veightz/sensebook/dict/LocalDict.kt
package com.veightz.sensebook.dict

import android.content.Context
import org.json.JSONObject
import java.util.Locale
import java.util.concurrent.atomic.AtomicReference

/**
 * Bundled EN→ZH dict (same JSON shape as userscript/dict).
 * Asset: assets/dict/en-zh-common.json (half10k trial copy for MVP size).
 */
class LocalDict private constructor(
    private val version: String,
    private val count: Int,
    private val entries: Map<String, Entry>
) {
    data class Entry(
        val gloss: String,
        val pos: String? = null,
        /** Not present in current ECDICT subset; reserved for easy show-if-available. */
        val phonetic: String? = null
    )

    data class LookupResult(
        val hit: Entry?,
        val dictVersion: String,
        val dictCount: Int
    )

    fun lookup(raw: String): LookupResult {
        val key = normalize(raw) ?: return LookupResult(null, version, count)
        val hit = entries[key]
            ?: stemLookup(key)
        return LookupResult(hit, version, count)
    }

    private fun stemLookup(key: String): Entry? {
        // Light stems aligned with userscript: -s / -ed / -ing
        val candidates = buildList {
            if (key.endsWith("ies") && key.length > 4) add(key.dropLast(3) + "y")
            if (key.endsWith("es") && key.length > 3) add(key.dropLast(2))
            if (key.endsWith("s") && !key.endsWith("ss") && key.length > 2) add(key.dropLast(1))
            if (key.endsWith("ed") && key.length > 3) {
                add(key.dropLast(2))
                if (key.length > 4 && key[key.length - 3] == key[key.length - 4]) {
                    add(key.dropLast(3)) // stopped → stop
                }
            }
            if (key.endsWith("ing") && key.length > 4) {
                add(key.dropLast(3))
                add(key.dropLast(3) + "e")
            }
        }
        for (c in candidates) {
            entries[c]?.let { return it }
        }
        return null
    }

    companion object {
        private const val ASSET_PATH = "dict/en-zh-common.json"
        private val cached = AtomicReference<LocalDict?>(null)

        fun get(context: Context): LocalDict {
            cached.get()?.let { return it }
            synchronized(this) {
                cached.get()?.let { return it }
                val loaded = load(context.applicationContext)
                cached.set(loaded)
                return loaded
            }
        }

        private fun load(context: Context): LocalDict {
            context.assets.open(ASSET_PATH).bufferedReader().use { reader ->
                val root = JSONObject(reader.readText())
                val version = root.optString("version", "unknown")
                val count = root.optInt("count", 0)
                val entriesJson = root.getJSONObject("entries")
                val map = HashMap<String, Entry>(entriesJson.length())
                val keys = entriesJson.keys()
                while (keys.hasNext()) {
                    val k = keys.next()
                    val obj = entriesJson.getJSONObject(k)
                    val gloss = obj.optString("g", "")
                    if (gloss.isBlank()) continue
                    val pos = obj.optString("p").takeIf { it.isNotBlank() }
                    val phonetic = obj.optString("ph").takeIf { it.isNotBlank() }
                        ?: obj.optString("phonetic").takeIf { it.isNotBlank() }
                    map[k.lowercase(Locale.ROOT)] = Entry(gloss, pos, phonetic)
                }
                return LocalDict(version, if (count > 0) count else map.size, map)
            }
        }

        /** Single Latin token, length ≤ 20 — same short-word gate as userscript. */
        fun normalize(raw: String): String? {
            val t = raw.trim().lowercase(Locale.ROOT)
            if (t.isEmpty() || t.length > 20 || t.any { it.isWhitespace() }) return null
            if (!t.all { it in 'a'..'z' || it == '\'' || it == '-' }) return null
            return t.trim('\'', '-')
        }
    }
}

File: macos/Sensebook/Dict/LocalDict.swift
import Foundation

/// Bundled EN→ZH dict (same JSON shape as userscript/dict & Android assets).
final class LocalDict {
    static let shared: LocalDict = {
        LocalDict.loadBundled()
    }()

    struct Entry {
        let gloss: String
        let pos: String?
        let phonetic: String?
    }

    struct LookupHit {
        let entry: Entry
        let dictVersion: String
        let dictCount: Int
    }

    private let version: String
    private let count: Int
    private let entries: [String: Entry]

    private init(version: String, count: Int, entries: [String: Entry]) {
        self.version = version
        self.count = count
        self.entries = entries
    }

    var dictVersion: String { version }
    var dictCount: Int { count }

    func lookup(_ raw: String) -> LookupHit? {
        guard let key = Self.normalize(raw) else { return nil }
        if let hit = entries[key] ?? stemLookup(key) {
            return LookupHit(entry: hit, dictVersion: version, dictCount: count)
        }
        return nil
    }

    private func stemLookup(_ key: String) -> Entry? {
        var candidates: [String] = []
        if key.hasSuffix("ies"), key.count > 4 {
            candidates.append(String(key.dropLast(3)) + "y")
        }
        if key.hasSuffix("es"), key.count > 3 {
            candidates.append(String(key.dropLast(2)))
        }
        if key.hasSuffix("s"), !key.hasSuffix("ss"), key.count > 2 {
            candidates.append(String(key.dropLast(1)))
        }
        if key.hasSuffix("ed"), key.count > 3 {
            candidates.append(String(key.dropLast(2)))
            if key.count > 4 {
                let chars = Array(key)
                if chars[chars.count - 3] == chars[chars.count - 4] {
                    candidates.append(String(key.dropLast(3)))
                }
            }
        }
        if key.hasSuffix("ing"), key.count > 4 {
            candidates.append(String(key.dropLast(3)))
            candidates.append(String(key.dropLast(3)) + "e")
        }
        for c in candidates {
            if let e = entries[c] { return e }
        }
        return nil
    }

    /// Single Latin token, length ≤ 20 — same short-word gate as userscript / Android.
    static func isShortWord(_ raw: String) -> Bool {
        normalize(raw) != nil
    }

    static func normalize(_ raw: String) -> String? {
        let t = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if t.isEmpty || t.count > 20 { return nil }
        if t.contains(where: { $0.isWhitespace }) { return nil }
        guard t.allSatisfy({ ($0 >= "a" && $0 <= "z") || $0 == "'" || $0 == "-" }) else { return nil }
        return t.trimmingCharacters(in: CharacterSet(charactersIn: "'-"))
    }

    private static func loadBundled() -> LocalDict {
        let url =
            Bundle.main.url(forResource: "en-zh-common", withExtension: "json", subdirectory: "dict")
            ?? Bundle.main.url(forResource: "en-zh-common", withExtension: "json")
        guard let url,
              let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return LocalDict(version: "missing", count: 0, entries: [:])
        }
        let version = root["version"] as? String ?? "unknown"
        let count = root["count"] as? Int ?? 0
        let entriesJson = root["entries"] as? [String: Any] ?? [:]
        var map: [String: Entry] = [:]
        map.reserveCapacity(entriesJson.count)
        for (k, v) in entriesJson {
            guard let obj = v as? [String: Any] else { continue }
            let gloss = (obj["g"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !gloss.isEmpty else { continue }
            let pos = (obj["p"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            let phonetic = ((obj["ph"] as? String) ?? (obj["phonetic"] as? String))?
                .trimmingCharacters(in: .whitespacesAndNewlines)
            map[k.lowercased()] = Entry(
                gloss: gloss,
                pos: (pos?.isEmpty == false) ? pos : nil,
                phonetic: (phonetic?.isEmpty == false) ? phonetic : nil
            )
        }
        return LocalDict(version: version, count: count > 0 ? count : map.count, entries: map)
    }
}
